src/main/scala/cosas/fns/functions.scala

```diff
@@ -1,61 +1,43 @@
 package ohnosequences.cosas.fns
 
-trait AnyFn { type Out }
-
-trait OutBound[X] extends AnyFn { type Out <: X }
-trait Out[X]      extends AnyFn { type Out = X }
-
-trait ContainerOut extends AnyFn { type O }
-trait OutInContainer[F[_]] extends ContainerOut { type Out = F[O] }
-trait InContainer[X] extends ContainerOut { type O = X }
-
-
-/* Fns with different number of arguments */
-trait AnyFn0 extends AnyFn with shapeless.DepFn0
-
-trait Fn0 extends AnyFn0
-
-
-trait AnyFn1 extends AnyFn {
+/* Dependent functions aka dependent products */
+trait AnyDepFn {
+  type Out
+}
+trait AnyDepFn0 extends AnyDepFn
+trait DepFn0[O] extends AnyDepFn0 {
```

Comment:
I think better not to have name clashes with shapeless. Let's name them just `FnX` or `DepFunX`?